sopel/test_tools.py

```diff
@@ -74,6 +75,12 @@ def _init_config(self):
             os.mkdir(home_dir)
         cfg.parser.set('core', 'homedir', home_dir)
 
+    def search_url_callbacks(self, url):
```

Comment:
I honestly don't know why. :-/

Reply:
Ok, I totally need to add the register/unregister method too, because it might fail in the future.

It's complicated, but TL;DR: it works now because none of the plugin that use register/unregister have test example, so their setup function is never called, so these methods are never called. However, the URL module is tested, and it uses the search method, so it needs to be here.

So, right now, it's pure luck that it works. I'll fix that and make sure it works *because that's the right thing to do*.

Reply:
OK, now we're adding three of `bot`'s methods to the `MockSopel` but not any of the others that modules might call. Something feels very wrong about this.

`say`, `reply`, and `action` are defined in `MockSopelWrapper` to store the command's output:

https://github.com/sopel-irc/sopel/blob/d2e7a163ff06e7697e857e77b14681cac251c845/sopel/test_tools.py#L87

We're not already looking at test problems because, for example, none of the core modules that use `bot.db` have testable examples. If they did have example results, errors would happen.

Same for `notice`—no core module command both has a testable example and calls `bot.notice()` while doing its work. If that happened, error.

I'm not going to "Unresolve" this conversation, because doing anything about any of this would be way out of scope for this PR, but… Basically, next time you feel ambitious about making the _tests_ better, it would be worth finding an alternative solution, like using the real Sopel but mocking out only its connection to IRC instead. Keeping the "real" and "mock" implementations/APIs in sync is only going to get sillier as we add features to the API and try to improve test coverage of modules.

Reply:
I fully agree with you. I'm not happy with these mocks, and they don't cover everything we need.

As said on IRC, we need time to figure something out that will be better and fit our needs. I'm sure we just need some time for this to sink in and come back to that topic with fresh ideas.